packages/peer-exchange/src/waku_peer_exchange.ts

```diff
@@ -72,45 +57,43 @@ export class WakuPeerExchange implements IPeerExchange {
 
     const stream = await this.newStream(peer);
 
-    await pipe(
+    const res = await pipe(
       [rpcQuery.encode()],
       lp.encode(),
       stream,
       lp.decode(),
       async (source) => await all(source)
     );
-  }
 
-  /**
-   * Handle a peer exchange query response
-   */
-  private handler(streamData: IncomingStreamData): void {
-    const { stream } = streamData;
-    pipe(stream, lp.decode(), async (source) => {
-      for await (const bytes of source) {
-        const decoded = PeerExchangeRPC.decode(bytes).response;
-
-        if (!decoded) {
-          throw new Error("Failed to decode response");
-        }
-
-        const enrs = await Promise.all(
-          decoded.peerInfos.map(
-            (peerInfo) => peerInfo.enr && ENR.decode(peerInfo.enr)
-          )
-        );
-
-        const peerInfos = enrs.map((enr) => {
-          return {
-            ENR: enr,
-          };
-        });
-
-        if (!this.callback) throw new Error("Callback not set");
-
-        await this.callback({ peerInfos });
+    try {
+      const bytes = new Uint8ArrayList();
+      res.forEach((chunk) => {
+        bytes.append(chunk);
+      });
+
+      const decoded = PeerExchangeRPC.decode(bytes).response;
+
+      if (!decoded) {
+        throw new Error("Failed to decode response");
       }
-    }).catch((err) => log("Failed to handle peer exchange request", err));
+
+      const enrs = await Promise.all(
+        decoded.peerInfos.map(
+          (peerInfo) => peerInfo.enr && ENR.decode(peerInfo.enr)
+        )
+      );
+
+      const peerInfos = enrs.map((enr) => {
+        return {
+          ENR: enr,
+        };
+      });
+
+      return peerInfos;
+    } catch (err) {
+      log("Failed to decode push reply", err);
+      throw new Error("Failed to decode push reply");
```

Comment:
In my opinion we should try, if possible, to make error messages consistent so that if it was reported to logs then it should be used in thrown error as well.

Reply:
We should either log and not throw or throw and not log. The redundancy is not necessary.
If we want to throw, then indeed, we should throw with the `err` so it's not lost in the stack.